grpc-server-security-jwt/src/main/java/io/micronaut/grpc/server/security/jwt/interceptor/GrpcServerSecurityJwtInterceptor.java

```diff
@@ -0,0 +1,209 @@
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.grpc.server.security.jwt.GrpcServerSecurityJwtConfiguration;
+import io.micronaut.security.config.InterceptUrlMapPattern;
+import io.micronaut.security.rules.SecurityRule;
+import io.micronaut.security.token.RolesFinder;
+import io.micronaut.security.token.jwt.generator.claims.JwtClaimsSetAdapter;
+import io.micronaut.security.token.jwt.validator.JwtValidator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.ParseException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+
+/**
+ * gRPC Server Security JWT Interceptor.
+ *
+ * @since 2.4.0
+ * @author Brian Wyka
+ */
+public class GrpcServerSecurityJwtInterceptor implements ServerInterceptor, Ordered {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GrpcServerSecurityJwtInterceptor.class);
+
+    private final GrpcServerSecurityJwtConfiguration config;
+    private final Metadata.Key<String> jwtMetadataKey;
+    private final JwtValidator jwtValidator;
+    private final boolean rejectRolesNotFound;
+    private final RolesFinder rolesFinder;
+
+    /**
+     * Create the interceptor based on the configuration.
+     *
+     * @param config the gRPC Security JWT configuration
+     * @param jwtValidator the JWT validator
+     * @param rolesFinder the roles finder
+     * @param rejectRolesNotFound whether or not to reject request if no roles found
+     */
+    public GrpcServerSecurityJwtInterceptor(final GrpcServerSecurityJwtConfiguration config,
+                                            final JwtValidator jwtValidator,
+                                            final RolesFinder rolesFinder,
+                                            final boolean rejectRolesNotFound) {
+        this.config = config;
+        this.jwtMetadataKey = Metadata.Key.of(config.getMetadataKeyName(), Metadata.ASCII_STRING_MARSHALLER);
+        this.jwtValidator = jwtValidator;
+        this.rejectRolesNotFound = rejectRolesNotFound;
+        this.rolesFinder = rolesFinder;
+    }
+
+    /**
+     * Intercept the call to validate the JSON web token.  If the token is not present in the metadata, or
+     * if the token is not valid, this method will deny the request with a {@link io.grpc.StatusRuntimeException}.
+     *
+     * @param call the server call
+     * @param metadata the metadata
+     * @param next the next processor in the interceptor chain
+     * @param <T> the type of the server request
+     * @param <S> the type of the server response
+     * @throws io.grpc.StatusRuntimeException if token not present or invalid
+     */
+    @Override
+    public <T, S> ServerCall.Listener<T> interceptCall(final ServerCall<T, S> call, final Metadata metadata, final ServerCallHandler<T, S> next) {
+        final List<String> requiredAccesses = getRequiredAccesses(config, call);
```

Comment:
Why is this not also using the SecurityRule API?

Reply:
`SecurityRule` implies knowledge of `HttpRequest` and `RouteMatch`:
https://github.com/micronaut-projects/micronaut-security/blob/master/security/src/main/java/io/micronaut/security/rules/SecurityRule.java#L57

I don't have either of those in this context.

I was attempting to use the `@Secured` annotation, but failed since I didn't have `RouteMatch` accessible...

Reply:
@brianwyka OK makes sense. In that case I think a new API to support the same use cases should be created. It should in general work the same as the existing logic

Reply:
Alright, I'll take a look into that.

Reply:
That is what I was thinking, @jameskleeh maybe there should be some `MethodSecurityRule` interface

Reply:
Is there any way to get access to the potential `@Secured` annotation on the method as well?

Reply:
I think I can model after the `SecurityFilter` approach.  Create one high-level `GrpcServerSecurityInterceptor` and have all the method rules collected there and interrogated.  Sound reasonable?

Reply:
I may open a PR on `micronaut-security` as I am realizing I end up duplicating a lot of code here already implemented in that project due to `HttpRequest`.  I'm thinking about abstracting some of the logic and making the security more pluggable and parameterized (generic in the java sense)

Reply:
I started to do some of the duplication in this repository... Here is the start of it: https://github.com/micronaut-projects/micronaut-grpc/pull/322/commits/7c676d0092cbef6b7cb8f432d980d72ddbe4fb3b.  I bypassed creating the generic token stuff and focused on just JWT for now.  Thoughts?